arrows/vxl/image_statistics.h

```diff
@@ -0,0 +1,142 @@
+    {
+      unsigned i = position % ni;
+      unsigned j = ( position / ni ) % nj;
+      dst.push_back( src( i, j, p ) );
+    }
+  }
+
+  std::sort( dst.begin(), dst.end() );
+
+  if( remove_extremes )
+  {
+    constexpr auto low = PixType{ 0 };
+
+    while( !dst.empty() && dst.front() == low )
+    {
+      dst.erase( dst.begin() );
+    }
+
+    constexpr auto high = std::numeric_limits< PixType >::max();
+
+    while( !dst.empty() && dst.back() == high )
+    {
+      dst.pop_back();
+    }
+  }
+  return dst;
+}
+
+// ----------------------------------------------------------------------------
+// Estimate the pixel values at given percentiles using a subset of points
+template < typename PixType >
+std::vector< PixType >
+get_image_percentiles(
+  vil_image_view< PixType > const& src,
+  std::vector< double > const& percentiles,
+  unsigned sampling_points, bool remove_extremes = false )
+{
+  std::vector< PixType > sorted_samples =
+    sample_and_sort_image( src, sampling_points, remove_extremes );
+
+  std::vector< PixType > dst( percentiles.size() );
+  double sampling_points_minus1 =
+    static_cast< double >( sorted_samples.size() - 1 );
+
+  for( auto const i : kwiver::vital::range::iota( percentiles.size() ) )
+  {
+    // Find the index by multiplying the number of points by the percentile
+    // The number is adjusted by -1 to account for the fact that percentiles
+    // are the number which fall below a value. The +0.5 is to account for
+    // truncation by static cast.
+    auto ind =
+      static_cast< size_t >(
+        sampling_points_minus1 * percentiles[ i ] + 0.5 );
+    dst[ i ] = sorted_samples[ ind ];
+  }
+  return dst;
+}
+
+// ----------------------------------------------------------------------------
+// Resultant image is true where pixel is greater than a percentile threshold
+template < typename PixType >
+void
+percentile_threshold_above(
+  vil_image_view< PixType > const& src,
+  std::vector< double > const& percentiles,
+  vil_image_view< bool >& dst, unsigned sampling_points = 1000 )
+{
+  // Calculate thresholds
+  auto thresholds = get_image_percentiles( src, percentiles, sampling_points );
+  dst.set_size( src.ni(), src.nj(),
+                static_cast< unsigned >( percentiles.size() ) );
+
+  // Perform thresholding
+  for( unsigned i = 0; i < thresholds.size(); i++ )
+  {
+    vil_image_view< bool > output_plane = vil_plane( dst, i );
+    vil_threshold_above( src, output_plane, thresholds[ i ] );
+  }
+}
+
+// ----------------------------------------------------------------------------
+// Resultant image is true where pixel is greater than a percentile threshold
+template < typename PixType >
+void
+percentile_threshold_above(
+  const vil_image_view< PixType >& src, const double percentile,
+  vil_image_view< bool >& dst, unsigned sampling_points = 1000 )
+{
+  percentile_threshold_above(
+    src, std::vector< double >( 1, percentile ), dst, sampling_points );
```

Comment:
```suggestion
    src, std::vector< double >{ 1, percentile }, dst, sampling_points );
```

Reply:
@mwoehlke-kitware here's another subtle example of uniform initialization going sideways. This is supposed to be a one-length vector containing `percentile` but it looks like on g++ 7.5.0 it treats it as a two-length vector.

Reply:
ah, yes, the cursed `vector` ctor strikes again...

BTW, `std::vector< double >{ percentile }` also works.